### DISCUSSION TITLE
Help required to simulate working of scrcpy-server.jar

### DISCUSSION THREAD

**Discussion by krishtoautomate** [NetrisTV/ws-scrcpy#131] [Show and tell]
Can you please help me capture buffer data from scrcpy-server.jar

**Step 1 :** 

adb push dist/server/vendor/Genymobile/scrcpy/scrcpy-server.jar /data/local/tmp

**Step 2:**

adb shell 'CLASSPATH=/data/local/tmp/scrcpy-server.jar app_process / com.genymobile.scrcpy.Server 1.17-ws5 DEBUG 0 8000000 60 -1 false - false true 0 false false - web 8886' 

**error:**

[server] ERROR: Exception on thread Thread[main,5,main]
java.lang.IllegalArgumentException: Expecting 15 parameters
	at com.genymobile.scrcpy.Server.parseArguments(Server.java:43)
	at com.genymobile.scrcpy.Server.main(Server.java:152)
	at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
	at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:409)

**Step 3:**

adb forward tcp:8666 localabstract:scrcpy

**Step 4:** 

nc localhost 8666

**Comment by drauggres** [comment]
***Step 2:**
> adb shell 'CLASSPATH=/data/local/tmp/scrcpy-server.jar app_process / com.genymobile.scrcpy.Server 1.17-ws5 web DEBUG 8886'

***Step 3:**
> `adb forward tcp:8886 tcp:8886`

Or you can skip it and connect to `DEVICE_IP:8886`  directly, `scrcpy-server.jar` from this project listens on TCP port, not a `localabstract`.

But you will not see anything in `nc localhost 8666`, you have a Web Socket server on the other end.

What exactly are you trying to achieve?

**Comment by krishtoautomate** [comment]
i want to create similar to below project for ws-scrcpy

https://github.com/DeviceFarmer/minicap

run jar in terminal(command line)
capture and display video in url.

ws-scrcpy is too big to understand.

**Comment by krishtoautomate** [comment]
Still dont receive any data:

**step 1:**
adb shell 'CLASSPATH=/data/local/tmp/scrcpy-server.jar app_process / com.genymobile.scrcpy.Server 1.17-ws5 web DEBUG 8886'

**Response:** [server] DEBUG: Server started! ::/:::8886

**step 2:** 
const WebSocket = require('ws');

var webSocket = new WebSocket('ws://127.0.0.1:8666');
webSocket.onmessage = function(data) { console.log(data); }


Response: error

Error: socket hang up
    at connResetException (internal/errors.js:570:14)
    at Socket.socketOnEnd (_http_client.js:440:23)
    at Socket.emit (events.js:228:7)
    at endReadableNT (_stream_readable.js:1185:12)
    at processTicksAndRejections (internal/process/task_queues.js:81:21)
Emitted 'error' event on WebSocket instance at:
    at ClientRequest.emit (events.js:223:5)
    at Socket.socketOnEnd (_http_client.js:440:9)
    at Socket.emit (events.js:228:7)
    at endReadableNT (_stream_readable.js:1185:12)
    at processTicksAndRejections (internal/process/task_queues.js:81:21) {
  code: 'ECONNRESET'
}

**Comment by drauggres** [comment]
Server in `scrcpy-server.jar` from this project is a Web Socket Server (ws in `ws-scrcpy` stands for Web Socket).
To connect to WebSockets from node.js I use https://github.com/websockets/ws.
For browser you will need to do something like this:
```typescript
const ws = new WebSocket(`ws://127.0.0.1:8886/`); 
ws.binaryType = 'arraybuffer';
ws.onmessage =  function(data) { console.log(data); };
```

Code:
https://github.com/NetrisTV/ws-scrcpy/blob/67cc51c90861cadd1f977f14bc3efe8b5dfa56ca/src/app/client/ManagerClient.ts#L29-L37
https://github.com/NetrisTV/ws-scrcpy/blob/67cc51c90861cadd1f977f14bc3efe8b5dfa56ca/src/app/client/StreamReceiver.ts#L50-L51
 

Or you might want to use original `scrcpy-server.jar` and use raw TCP from node.js. To find arguments for it in this case:
1. run scrcpy
2. find server pid, something like: `adb shell 'ps -A | grep app_process'`
3. find arguments: `adb shell 'cat /proc/<PID>/cmdline'`
Arguments will be separated by `\0`, i.e. look like there is no space.

**Comment by krishtoautomate** [comment]
Tried but no luck

**Step 1:**

adb shell 'CLASSPATH=/data/local/tmp/scrcpy-server.jar app_process / com.genymobile.scrcpy.Server 1.17-ws5 web DEBUG 8886'

Response: [server] DEBUG: Server started! ::/:::8886


**step 2:**

const WebSocket = require('ws');

const ws = new WebSocket(`ws://127.0.0.1:8886/`);
ws.binaryType = 'arraybuffer';
ws.onmessage =  function(data) { console.log(data); };

**error:**

events.js:200
      throw er; // Unhandled 'error' event
      ^

Error: connect ECONNREFUSED 127.0.0.1:8886
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1134:16)
Emitted 'error' event on WebSocket instance at:
    at ClientRequest.<anonymous> (/Users/user/node-scrcpy-client/node_modules/ws/lib/websocket.js:579:15)
    at ClientRequest.emit (events.js:223:5)
    at Socket.socketErrorListener (_http_client.js:406:9)
    at Socket.emit (events.js:223:5)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:81:21) {
  errno: 'ECONNREFUSED',
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 8886
}

**Comment by drauggres** [comment]
You must either forward port to localhost (`127.0.0.1`) with 
```shell
adb forward tcp:8886 tcp:8886
```
 or connect to device directly
```typescript
const ip = '192.168.1.101';   // change to your ip address here
const ws = new WebSocket(`ws://${ip}:8886/`)
```

**Comment by krishtoautomate** [comment]
adb shell 'killall app_process'
adb push dist/server/vendor/Genymobile/scrcpy/scrcpy-server.jar /data/local/tmp/scrcpy-server.jar

adb shell 'CLASSPATH=/data/local/tmp/scrcpy-server.jar app_process / com.genymobile.scrcpy.Server 1.17-ws5 web DEBUG 8886'

adb forward tcp:8886 tcp:8886

const WebSocket = require('ws');

const ws = new WebSocket(`ws://localhost:8886/`);

ws.binaryType = 'arraybuffer';
ws.onmessage =  function(data) { console.log(data); };


**worked.**

**Comment by krishtoautomate** [comment]
why buffer is not continues?

how do i display to browser?

**Comment by krishtoautomate** [comment]
should i send any message to websocket to receive video buffer array in non-stop?

**Comment by drauggres** [comment]
After connection you should receive initial info:
https://github.com/NetrisTV/ws-scrcpy/blob/67cc51c90861cadd1f977f14bc3efe8b5dfa56ca/src/app/client/StreamReceiver.ts#L131-L135
To start stream you must send `ControlMessage.TYPE_CHANGE_STREAM_PARAMETERS` message:
https://github.com/NetrisTV/ws-scrcpy/blob/67cc51c90861cadd1f977f14bc3efe8b5dfa56ca/src/app/googDevice/client/StreamClientScrcpy.ts#L207-L211
